### PR TITLE
Added tox configuration for testing building of astropy with latest developer version of build-time dependencies

### PR DESCRIPTION
**Pull request by astrofrog**
xref #15495

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by astrofrog** [comment]
cc @pllim @mhvk @saimn

File: tox.ini [hunk 1]
```diff
@@ -121,6 +121,22 @@ pip_pre =
...
+skip_install = true
+changedir = .tmp/{envname}
+setenv =
+    PIP_EXTRA_INDEX_URL = https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
+deps =
+    numpy>=0.0.dev0
+    git+https://github.com/pypa/setuptools.git
+    git+https://github.com/pypa/setuptools_scm.git
+    git+https://github.com/cython/cython
+    git+https://github.com/astropy/extension-helpers
```
Comment (mhvk) [inline tox.ini:136]: Would it help to include `pyerfa` here too as `git+https`?

**Comment by mhvk** [comment]
The builddev run fails :disappointed:

File: tox.ini (hunk 1 above)
Comment (astrofrog) [reply tox.ini:136]: No, it is not a build-time dependency

**Comment by astrofrog** [comment]
@mhvk - I think the failure was due to a typo, let's see if it works now

File: tox.ini (hunk 1 above)
Comment (mhvk) [reply tox.ini:136]: OK, makes sense!

**Comment by mhvk** [comment]
Still a failure, but now unrelated to this PR: numpy introduced some extra ufuncs...

File: tox.ini [hunk 2]
```diff
@@ -121,6 +121,22 @@ pip_pre =
...
+[testenv:devbuild]
+skip_install = true
+changedir = .tmp/{envname}
+setenv =
+    PIP_EXTRA_INDEX_URL = https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
+deps =
+    numpy>=0.0.dev0
+    git+https://github.com/pypa/setuptools.git
+    git+https://github.com/pypa/setuptools_scm.git
+    git+https://github.com/cython/cython
```
Comment (pllim) [inline tox.ini:135]: Are these really necessary? I thought we only have to build against numpy dev...

Comment (astrofrog) [reply tox.ini:135]: Well these are all build-time dependencies, right? Might as well know if one of them starts becoming incompatible?

**Comment by astrofrog** [comment]
@mhvk - I've added a commit to make sure that we actually run the tests against stable numpy. That way, any failures should truly be from the build phase in terms of numpy dev.

**Comment by mhvk** [comment]
Yes, that's a good idea - nice separation of concerns! One comment only: I think this should be done as part of `ci_cron_weekly` (or maybe daily), not on every PR/commit. After all, almost all PRs only affect astropy python code, so are extremely unlikely to cause failures.

**Review by pllim — COMMENTED** [review]
I am not sure that this solves the actual problem of us not *shipping* dev wheel using dev numpy?

File: tox.ini [hunk 3]
```diff
@@ -121,6 +121,24 @@ pip_pre =
...
 
+# Test that we can build astropy with the latest developer versions of all
+# build-time dependencies,
+[testenv:devbuild]
+skip_install = true
+changedir = .tmp/{envname}
+setenv =
+    PIP_EXTRA_INDEX_URL = https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
+deps =
+    numpy>=0.0.dev0
```
Comment (pllim) [inline tox.ini:132]: Don't we need to unpin `numpy<2` like https://github.com/astropy/astropy/pull/15495 ?

**Comment by astrofrog** [comment]
> I am not sure that this solves the actual problem of us not shipping dev wheel using dev numpy?

No it doesn't and that might be harder to do. Do we really need to though? What is the benefit of doing this?

File: tox.ini (hunk 3 above)
Comment (astrofrog) [reply tox.ini:132]: No because the ``--no-build-isolation`` flag means that ``pyproject.toml`` is ignored for build dependencies.

**Comment by pllim** [comment]
> Do we really need to though? What is the benefit of doing this?

Because numpy 2.0.dev wants to throw error when it finds C-extension importing old `numpy.core` stuff. And it will throw error if we *build* with something not numpy 2.0.dev. At least that's how I understood the problem.

**Comment by astrofrog** [comment]
Huh that seems strange - is numpy.core even relevant for the build phase?

**Comment by seberg** [comment]
The point is that any ABI change in NumPy requires me to force you to compile against NumPy 2.0.  Until now, all changes have been so small that it effectively didn't matter (yes you can run into undefined behavior/crashes, but very few or no-one will be using those things).

But, eventually, there are some things that would be nice to do that break the promise that things might work "well enough if you don't worry about the undefined behavior".  Maybe I can try to avoid that like hell, but overall even enough removals would make me feel better with just forcing the recompile (yes, I know this hurts adoption speed :().

So, building your nightlies against our nightlies would be much appreciated.  One hack I figured out (waiting for someone to point out how horrible it is) is that if you don't want to use `--pre`/`PIP_PRE=1` in some form (without version pin), to use `numpy<=2.0.0.dev0` which _explicitly_ allows our dev wheels.
In either case, you clearly need to customize the nightly wheel build by at least adding something like:
```
CIBW_ENVIRONMENT="PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple"
```
not sure if that can go into astropy or `OpenAstronomy/github-actions-workflows/.github/workflows/publish.yml`.  I *think* cibuildwheel doesn't use build isolation, so it might work to alternatively explicitly install the pre-release NumPy also.

(I am sorry, I am fighting trying to understand how to do deal with this, but while I figured out how this allows us to modify C-API meaningfully, etc.  this building problems is new to me.)

**Comment by mhvk** [comment]
@seberg - yes, that makes sense!  It has to be possible to break the ABI.

And I think @astrofrog's PR here is a good one in that it tests one critical aspect of it all: that something compiled with numpy 2.0 will work with when run with older versions of numpy. I guess it may als help with something @pllim noted, that we need to start producing numpy-2.0 compatible wheels so that downstream packages can start testing against numpy-2.0dev too. In principle, I guess those could be produced in this PR?

The other piece of the puzzle is that we need to make sure that in our regular numpy-dev test run, we also build with numpy-dev, and ensure that, somehow, our dependencies are built with numpy-dev as well since otherwise one gets failures. Here, again, could this PR help in principle? It already does the compilation, but only tests with older numpy now. Should it test against newer numpy too?

**Comment by astrofrog** [comment]
One thing that's not clear to me: does this means we may have to distribute two wheels for each Python version and architecture, one compiled with numpy<2 and one with numpy 2?

**Comment by mhvk** [comment]
I think we can have just one wheel for numpy 2, no? For most recent python and amd64?

(And perhaps we should see if we can move the limited python api like you did for erfa?)

**Comment by astrofrog** [comment]
Sorry I meant in the long term, once numpy 2.0 is out, do we have to distribute two of every wheel?

**Comment by seberg** [comment]
You need a single wheel build against 2.x.  Just right now if you cut a release, that would be weird to do, since we are in pre-release state and might get something wrong.
Everything else is NumPy's job to get right.  It isn't viable to ask you to have two wheels after all!

**Comment by astrofrog** [comment]
@seberg - ok so just to make sure I understand correctly:

* Wheels built against Numpy 1.x won't necessarily work at runtime with Numpy 2.x
* Wheels built against Numpy 2.x will work at runtime with both Numpy 1.x (for recent versions) and Numpy 2.x

Is this correct? If so, I agree that we need to therefore figure out how to build our dev wheels against dev numpy in the cibuildwheel builds, which I can investigate.

**Comment by seberg** [comment]
Yes, correct.  As I said elsehwere, I think cibuildwheel doesn't use build isolation so explicitly installing the newer NumPy may be a way to do it.
One caveat, I think it would be good to keep using the old NumPy version for testing the nightlies before upload.
(That is both great for NumPy, because it helps ensure we don't break that accidentally, and it makes sense for you, because you don't want nightly upload to fail on some minor change in `main`.)

**Comment by seberg** [comment]
Awesome, thanks!

> I'm not sure if we need to keep the present PR - it does test against the dev versions of all packages, but maybe that's overkill?

Maybe overkill, but I think you need some derivation of this to test against NumPy dev in CI, since that is distinct from the wheel building.

**Comment by astrofrog** [comment]
@seberg - for us the wheel building does happen in our regular CI as a nightly cron, so it's arguably part of the CI.

@pllim @mhvk - what do you think?

**Comment by pllim** [comment]
1. How painful is it to *build* astropy in just the devdeps job with numpy-dev but not touch the other jobs in CI workflow?
2. How painful is it to *build* astropy *nightly wheel* with numpy-dev and then upload it for downstream use?

**Comment by pllim** [comment]
Personally, I do not find this PR particularly useful in solving our predicament.

**Comment by pllim** [comment]
I don't think we need this anymore?

**Comment by astrofrog** [comment]
Let's just close this, the nightly dev wheels are fine.

**Pull request closed without merging**